### PR TITLE
checkpoint/restore: make sure the donated stdioFDs have the same value

### LINKED ISSUE #1844: Restore with --platform kvm failed
https://groups.google.com/forum/#!topic/gvisor-users/iNRqSJArWs4

I'm experimenting with gvisor's checkpoint/restore functionality.
With the default ptrace platform, it works well using a "bash" rootfs
but with kvm platform, I got the below panic on restore.

$sudo runsc restore --image-path=/home/aaron/dump/ bash
running container: starting container: restoring container "bash": MultiDevice's caches are inconsistent:
state path: *(*(*(*(*(*(*(*(*(*(*(*(*(*(root.globalInit).mounts).mounts.(key 5)).Inode).InodeOperations.(ramfs.Dir)).children["devices"]).InodeOperations.(ramfs.Dir)).children["system"]).InodeOperations.(ramfs.Dir)).children["cpu"]).InodeOperations.(ramfs.Dir)).children["cpu4"]).InodeOperations.(ramfs.Dir)).dentryMap).entries
goroutine 5 [running]:
gvisor.dev/gvisor/pkg/state.(*recoverable).safely.func1(0xc000188160, 0xc00013d640)
        pkg/state/state.go:347 +0x149
panic(0xc45a80, 0xeda2a0)
        GOROOT/src/runtime/panic.go:679 +0x1b2
gvisor.dev/gvisor/pkg/sentry/device.(*MultiDevice).Load(0xc0001f6360, 0x15, 0x0, 0x0, 0x4, 0x1, 0x4f53b2f34a65d600)
        pkg/sentry/device/device.go:256 +0x447
gvisor.dev/gvisor/pkg/sentry/fs/host.(*inodeFileState).afterLoad(0xc0001f8540)
        pkg/sentry/fs/host/inode_state.go:56 +0x127
gvisor.dev/gvisor/pkg/state.(*objectState).checkComplete(0xc0005a6a00, 0xc0002cb340)
        pkg/state/decode.go:86 +0x86
gvisor.dev/gvisor/pkg/state.(*objectState).checkComplete(0xc0005dbb00, 0xc0002cb340)
        pkg/state/decode.go:93 +0xf5
gvisor.dev/gvisor/pkg/state.(*decodeState).Deserialize(0xc000188120, 0xdbf940, 0xc000282820, 0x199)
        pkg/state/decode.go:523 +0x3d1
gvisor.dev/gvisor/pkg/state.Load.func1()
        pkg/state/state.go:118 +0xce
gvisor.dev/gvisor/pkg/state.(*recoverable).safely(0xc000188160, 0xc000251660, 0x0, 0x0)
        pkg/state/state.go:360 +0x6d
gvisor.dev/gvisor/pkg/state.Load(0x7f3ebdf4eac0, 0xc0000f1c80, 0xeebfe0, 0xc0003380e0, 0xdbefc0, 0xc000282820, 0xc0002cb340, 0x0, 0x0)
        pkg/state/state.go:117 +0xfc
gvisor.dev/gvisor/pkg/sentry/kernel.(*Kernel).LoadFrom(0xc000282820, 0xeebfe0, 0xc0003380e0, 0xf0e3c0, 0xc0000f6918, 0xef2500, 0xc0002f93e0, 0xc00030f9e0, 0x0)
        pkg/sentry/kernel/kernel.go:563 +0x426
gvisor.dev/gvisor/pkg/sentry/state.LoadOpts.Load(0xeed440, 0xc0000f6cc8, 0x0, 0x0, 0x0, 0xc000282820, 0xf0e3c0, 0xc0000f6918, 0xef2500, 0xc0002f93e0, ...)
        pkg/sentry/state/state.go:118 +0xe1
gvisor.dev/gvisor/runsc/boot.(*containerManager).Restore(0xc0002e4500, 0xc00030ebd0, 0x1514270, 0x0, 0x0)
        runsc/boot/controller.go:380 +0x7ac
reflect.Value.call(0xc000188900, 0xc0000f6a20, 0x13, 0xdc5275, 0x4, 0xc000251e88, 0x3, 0x3, 0xc000251cc8, 0x4bccad, ...)
        GOROOT/src/reflect/value.go:460 +0x5f6
reflect.Value.Call(0xc000188900, 0xc0000f6a20, 0x13, 0xc000251e88, 0x3,0x3, 0x0, xc0002e4500,0x16)
        GOROOT/src/reflect/value.go:321 +0xb4
gvisor.dev/gvisor/pkg/urpc.(*Server).handleOne(0xc00026be60, 0xc00031e000, 0x0, 0x0)
        pkg/urpc/urpc.go:325 +0x56d
gvisor.dev/gvisor/pkg/urpc.(*Server).handleRegistered(0xc00026be60,0xc00031e000, 0x0, 0x0)
        pkg/urpc/urpc.go:420 +0x35
gvisor.dev/gvisor/pkg/urpc.(*Server).StartHandling.func1(0xc00026be60,0xc00031e000)
        pkg/urpc/urpc.go:440 +0x73
created by gvisor.dev/gvisor/pkg/urpc.(*Server).StartHandling
        pkg/urpc/urpc.go:438 +0x61

$runsc --version
runsc version release-20200127.0-99-g17b9f5e66238-dirty
spec: 1.0.1-dev

Then I realized on restore, I also have to specify the same platform setting used for running the sandbox, i.e. for a checkpointed image that is generated from sandbox which uses ptrace platform, the image can only be restored to a new sandbox using ptrace platform. The same is true for kvm.

So, the following cmdline works:
$sudo runsc --platform kvm restore --image-path=/home/aaron/dump/ bash

This issue is created to improve the error message when invalid platform is used on restore.

### PR DESCRIPTION
**Pull request by aaronlu**
Suppose I start a runsc container using kvm platform like this:
$ sudo runsc --debug=true --debug-log=1.txt --platform=kvm run rootbash
The donating FD and the corresponding cmdline for runsc-sandbox is:

D0313 17:50:12.608203   44389 x:0] Donating FD 3: "1.txt"
D0313 17:50:12.608214   44389 x:0] Donating FD 4: "control_server_socket"
D0313 17:50:12.608224   44389 x:0] Donating FD 5: "|0"
D0313 17:50:12.608229   44389 x:0] Donating FD 6: "/home/ziqian.lzq/bundle/bash/runsc/config.json"
D0313 17:50:12.608234   44389 x:0] Donating FD 7: "|1"
D0313 17:50:12.608238   44389 x:0] Donating FD 8: "sandbox IO FD"
D0313 17:50:12.608242   44389 x:0] Donating FD 9: "/dev/kvm"
D0313 17:50:12.608246   44389 x:0] Donating FD 10: "/dev/stdin"
D0313 17:50:12.608249   44389 x:0] Donating FD 11: "/dev/stdout"
D0313 17:50:12.608253   44389 x:0] Donating FD 12: "/dev/stderr"
D0313 17:50:12.608257   44389 x:0] Starting sandbox: /proc/self/exe
[runsc-sandbox --root=/run/containerd/runsc/default --debug=true --log=
--max-threads=256 --reclaim-period=5 --log-format=text --debug-log=1.txt
--debug-log-format=text --file-access=exclusive --overlay=false
--fsgofer-host-uds=false --network=sandbox --log-packets=false
--platform=kvm --strace=false --strace-syscalls=--strace-log-size=1024
--watchdog-action=Panic --panic-signal=-1 --profile=false --net-raw=true
--num-network-channels=1 --rootless=false --alsologtostderr=false
--ref-leak-mode=disabled --gso=true --software-gso=true
--overlayfs-stale-read=false --shared-volume= --debug-log-fd=3
--panic-signal=15 boot --bundle=/home/ziqian.lzq/bundle/bash/runsc
--controller-fd=4 --mounts-fd=5 --spec-fd=6 --start-sync-fd=7 --io-fds=8
--device-fd=9 --stdio-fds=10 --stdio-fds=11 --stdio-fds=12 --pidns=true
--setup-root --cpu-num 32 --total-memory 4294967296 rootbash]

Note stdioFDs starts from 10 with kvm platform and stderr's FD is 12.

If I restore a container from the checkpoint image which is derived
by checkpointing the above rootbash container, but either omit the
platform switch or specify to use ptrace platform explicitely:
$ sudo runsc --debug=true --debug-log=1.txt restore --image-path=some_path restored_rootbash

the donating FD and corresponding cmdline for runsc-sandbox is:

D0313 17:50:15.258632   44452 x:0] Donating FD 3: "1.txt"
D0313 17:50:15.258640   44452 x:0] Donating FD 4: "control_server_socket"
D0313 17:50:15.258645   44452 x:0] Donating FD 5: "|0"
D0313 17:50:15.258648   44452 x:0] Donating FD 6: "/home/ziqian.lzq/bundle/bash/runsc/config.json"
D0313 17:50:15.258653   44452 x:0] Donating FD 7: "|1"
D0313 17:50:15.258657   44452 x:0] Donating FD 8: "sandbox IO FD"
D0313 17:50:15.258661   44452 x:0] Donating FD 9: "/dev/stdin"
D0313 17:50:15.258675   44452 x:0] Donating FD 10: "/dev/stdout"
D0313 17:50:15.258680   44452 x:0] Donating FD 11: "/dev/stderr"
D0313 17:50:15.258684   44452 x:0] Starting sandbox: /proc/self/exe
[runsc-sandbox --root=/run/containerd/runsc/default --debug=true --log=
--max-threads=256 --reclaim-period=5 --log-format=text --debug-log=1.txt
--debug-log-format=text --file-access=exclusive --overlay=false
--fsgofer-host-uds=false --network=sandbox --log-packets=false
--platform=ptrace --strace=false --strace-syscalls=
--strace-log-size=1024 --watchdog-action=Panic --panic-signal=-1
--profile=false --net-raw=true --num-network-channels=1 --rootless=false
--alsologtostderr=false --ref-leak-mode=disabled --gso=true
--software-gso=true --overlayfs-stale-read=false --shared-volume=
--debug-log-fd=3 --panic-signal=15 boot
--bundle=/home/ziqian.lzq/bundle/bash/runsc --controller-fd=4
--mounts-fd=5 --spec-fd=6 --start-sync-fd=7 --io-fds=8 --stdio-fds=9
--stdio-fds=10 --stdio-fds=11 --setup-root --cpu-num 32 --total-memory
4294967296 restored_rootbash]

Note this time, stdioFDs starts from 9 and stderr's FD is 11(so the
saved host.descritor.origFD which is 12 for stderr is no longer valid).

For the three host FD based files, The s.Dev and s.Ino derived from
fstat(fd) shall all be the same and since the two fields are used
as device.MultiDeviceKey, the host.inodeFileState.sattr.InodeId which is
the value of MultiDevice.Map(MultiDeviceKey), shall also all be the same.
Note that for MultiDevice m, m.cache records the mapping of key to value
and m.rcache records the mapping of value to key. If same value doesn't
map to the same key, it will panic on restore.

Now that stderr's origFD 12 is no longer valid(it happens to be
/memfd:runsc-memory in my test on restore), the s.Dev and s.Ino derived
from fstat(fd=12) in host.inodeFileState.afterLoad() will neither be
correct. But its InodeID is still the same as saved, MultiDevice.Load()
will complain about the same value(InodeID) being mapped to different
keys (different from stdin and stdout's) and panic with: "MultiDevice's
caches are inconsistent".

Solve this problem by making sure stdioFDs for root container's init
task are always the same on initial start and on restore time, no matter
what cmdline user has used: debug log specified or not, platform changed
or not etc. shall not affect the ability to restore.

Fixes #1844.

### CODE REVIEW

File: runsc/boot/loader.go [hunk 1]
```diff
@@ -319,6 +322,17 @@ func New(args Args) (*Loader, error) {
 		return nil, fmt.Errorf("creating pod mount hints: %v", err)
 	}
 
+	var stdioFDs []int
+	newfd := startingStdioFD
+	for _, fd := range args.StdioFDs {
+		err := syscall.Dup3(fd, newfd, 0)
```
Comment (fvoznika) [inline runsc/boot/loader.go]: Please set O_CLOEXEC on the new FD to make sure it doesn't "leak" on exec.

File: runsc/boot/loader.go [hunk 2]
```diff
@@ -319,6 +322,17 @@ func New(args Args) (*Loader, error) {
...
+		if err != nil {
+			return nil, fmt.Errorf("dup3 of stdioFDs failed: %v", err)
+		}
+		stdioFDs = append(stdioFDs, newfd)
```
Comment (fvoznika) [inline runsc/boot/loader.go]: You can close `fd` here since it's not going to be used anymore.

File: runsc/boot/loader.go (hunk 1 above)
Comment (aaronlu) [reply runsc/boot/loader.go]: Will do, thanks for pointing this out.

File: runsc/boot/loader.go (hunk 2 above)
Comment (aaronlu) [reply runsc/boot/loader.go]: Ah right. I had the false impression that these FDs are passed through unix domain sockets through urpc but they are actually inherited through cmd.Start(). Will close them, thanks.

**Comment by aaronlu** [comment]
Hi Fabricio,
I'm new to github review and not sure what's the next step to do. Shall I close this PR and send a new one?
BTW, do I need to press the 'resolve conversion' button? Thanks.

**Comment by fvoznika** [comment]
You don't need to send another PR. Just make the edits, amend your commit and push it again. For the comments that you have addressed, you can click `Resolve conversation`. Otherwise you can keep it open if you have question or need to discuss further.

**Comment by aaronlu** [comment]
Thanks for the guide, they are very helpful.

**Comment by aaronlu** [comment]
While testing, I saw goroutine stack dump following below msg:
W0331 02:57:13.570585       1 log.go:338] Watchdog.Start() not called within 30s:
appeared in log some time after restore completed successfully. It didn't make the restore fail, the restored container can accept commands without any problem. I also confirm the error message is there with or without this fix patch so should be a separate issue, I'll take a look at this watchdog issue some time later.

**Comment by aaronlu** [comment]
v2:
 - set CLOEXEC in dup3() and close fd after dup3() as suggested by Fabricio.
 - move the close of l.stdioFDs from inside !l.restore block to outside,
   so that l.stdioFDs are closed both on initial start and on restore.
   We can close these stdioFDs because they are either dup()ed in
   createFDTable() during initial start or dup()ed in descriptor.initAfterLoad()
   and no longer useful/needed.

**Pull request merged**